### PR TITLE
[CUDA] Bounds-check segment_reduce_backward_kernel offsets to fix OOB write

### LINKED ISSUE #192452: [CUDA] illegal memory write on segment_reduce_backward_kernel
### 🐛 Describe the bug

```python
import torch


torch.cuda.init()
gpu = torch.device("cuda")
data = torch.ones((5,), dtype=torch.float64, device=gpu, requires_grad=True)
lengths = torch.tensor([1, 1, 1, 2], dtype=torch.int64, device=gpu)
lengths_alias = torch.from_dlpack(lengths)

output = torch.segment_reduce(
    data,
    "max",
    lengths=lengths,
    axis=0,
    unsafe=False,
    initial=1.0,
)
lengths_alias.fill_(-(2**63))
output.sum().backward()
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ write of size 8 bytes
=========     at void at::native::<unnamed>::segment_reduce_backward_kernel<double, long>(at::native::ReductionType, T1 *, const T1 *, const T1 *, const T1 *, const T2 *, const T2 *, long, long, T1, long, long, long, long, long, long, long)+0x3d60
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7f5814000000 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f5813e00000 of size 2097152 bytes
```

### Versions

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by PannenetsF**
192452 也 review 完了，lint后原样push加mr

Fixes #192452

> ## Root cause
>
> `segment_reduce_backward_kernel` derived its per-segment loop bounds
> (`offset_start`, `offset_end`) from the lengths cumsum buffer without
> validating them against the data extent. When the `lengths` tensor is mutated
> between the forward and backward passes -- the issue's repro aliases it via
> `torch.from_dlpack` and does `fill_(INT64_MIN)` -- the recomputed offsets fall
> far outside `[0, data_size_axis)`. The loop then writes `grad_input` at wild
> indices, producing an out-of-bounds global write that `compute-sanitizer`
> reports as `Invalid __global__ write of size 8 bytes`. The forward kernel
> already guards `lengths >= 0`, but the backward kernel had no equivalent check.
>
> ## Fix
>
> Add a `CUDA_KERNEL_ASSERT` in `segment_reduce_backward_kernel` bounding
> `segment_length`, `offset_start`, and `offset_end` against `data_size_axis`,
> and move the `segment_length == 0` early-return to after the offset reads.
> Reading the two offset entries before the early-return is always safe: the
> cumsum buffer has `segment_count + 1` entries along the axis and
> `dim_idx < segment_count`, so `offset_idx + 1` is in range.
>
> This mirrors the forward kernel's existing `lengths >= 0` guard rather than
> clamping the offsets. Clamping would mask the caller's bug and still produce
> wrong gradients, whereas the assert turns silent memory corruption on corrupted
> input into a deterministic device-side assert.
>
> ## Impact
>
> Valid inputs are unaffected. Corrupted or mutated lengths that previously
> caused an out-of-bounds write now fail fast with a device-side assert.
>
> A new `@onlyCUDA` test, `test_backward_rejects_mutated_lengths`, runs the
> issue's DLPack-alias repro in a subprocess and checks that the CUDA/HIP failure
> is surfaced in stderr.
>
> ## Validation
>
> ```
> lintrunner --skip CLANGTIDY,CLANGFORMAT,CLANGTIDY_EXECUTORCH_COMPATIBILITY \
>     aten/src/ATen/native/cuda/SegmentReduce.cu test/test_segment_reductions.py
> ```
>
> Result: `No lint issues.`
>
> ```
> git diff main...HEAD --check
> ```
>
> Result: no whitespace errors.
>
> CUDA runtime validation was unavailable on the local macOS/arm64 host. CI or a
> CUDA reviewer can run the added test and the issue's compute-sanitizer repro.
>
> ## AI-assistance disclosure
>
> This PR was authored with help from an AI assistant. All code and this
> description were reviewed by the author, who understands the change and takes
> responsibility for it.

### CODE REVIEW

File: aten/src/ATen/native/cuda/SegmentReduce.cu [hunk 1]
```diff
@@ -188,13 +188,15 @@ __global__ void segment_reduce_backward_kernel(
 
   int64_t lengths_idx = outer_idx * lengths_stride_axis * segment_count + dim_idx;
   auto segment_length = lengths_data[lengths_idx];
-  if (segment_length == 0) {
```
Comment (eqy) [inline aten/src/ATen/native/cuda/SegmentReduce.cu:191]: why is this branch removed? isn't it useful as a short-circuit path to save time?

**Comment by eqy** [comment]
@claude review this PR, paying close attention to bugs human reviewers may have missed and weighing the tradeoffs of additional overhead and improved user ergonomics/safety carefully

File: aten/src/ATen/native/cuda/SegmentReduce.cu (hunk 1 above)
Comment (PannenetsF) [reply aten/src/ATen/native/cuda/SegmentReduce.cu:191]: @eqy This was already addressed in b284e22 by restoring the zero-length short-circuit; sorry, I forgot to tag you earlier.
